core/src/main/java/com/linecorp/armeria/common/logging/RequestScopedMdc.java

```diff
@@ -0,0 +1,377 @@
+                tmp.remove(key);
+                newMap = Collections.unmodifiableMap(tmp);
+            }
+            ctx.setAttr(MAP, newMap);
+        }
+    }
+
+    /**
+     * Unbinds all request-scoped {@link MDC} properties from the specified {@link RequestContext}.
+     *
+     * @param ctx the {@link RequestContext}
+     */
+    public static void clear(RequestContext ctx) {
+        requireNonNull(ctx, "ctx");
+
+        synchronized (ctx) {
+            final Map<String, String> oldMap = getMap(ctx);
+            if (!oldMap.isEmpty()) {
+                ctx.setAttr(MAP, Collections.emptyMap());
+            }
+        }
+    }
+
+    private static Map<String, String> getMap(RequestContext ctx) {
+        final Map<String, String> map;
+        if (ctx instanceof ClientRequestContext) {
+            map = ((ClientRequestContext) ctx).ownAttr(MAP);
+        } else {
+            map = ctx.attr(MAP);
+        }
+        return firstNonNull(map, Collections.emptyMap());
+    }
+
+    private RequestScopedMdc() {}
+
+    private static final class Adapter implements MDCAdapter {
+
+        private final MDCAdapter delegate;
+
+        Adapter(MDCAdapter delegate) {
+            this.delegate = requireNonNull(delegate, "delegate");
+        }
+
+        @Override
+        @Nullable
+        public String get(String key) {
+            final RequestContext ctx = RequestContext.currentOrNull();
+            if (ctx != null) {
+                final String value = getMap(ctx).get(key);
+                if (value != null) {
+                    return value;
+                }
+            }
+
+            return delegate.get(key);
+        }
+
+        @Override
+        public Map<String, String> getCopyOfContextMap() {
+            final Map<String, String> threadLocalMap =
+                    firstNonNull(delegate.getCopyOfContextMap(), Collections.emptyMap());
+            final RequestContext ctx = RequestContext.currentOrNull();
+            if (ctx == null) {
+                // No context available
+                return threadLocalMap;
+            }
+
+            final Map<String, String> requestScopedMap =
+                    firstNonNull(getMap(ctx), Collections.emptyMap());
+            if (threadLocalMap.isEmpty()) {
+                // No thread-local map available
+                return requestScopedMap;
+            }
+
+            // Thread-local map available
+            if (requestScopedMap.isEmpty()) {
+                // Only thread-local map available
+                return threadLocalMap;
+            }
+
+            // Both thread-local and request-scoped map available
+            final Map<String, String> merged =
+                    new Object2ObjectOpenHashMap<>(threadLocalMap.size() + requestScopedMap.size());
+            merged.putAll(threadLocalMap);
+            merged.putAll(requestScopedMap);
+            return merged;
```

Comment:
Question: Is it intended to return the modifiable map?

Reply:
The API documentation does not tell us whether we should return an unmodifiable map or not, so I took a cheaper path.

Reply:
Yes, it seems so. Just wanted to make sure. 😉